### ISSUE TITLE
ModuleParseError: Module parse failed: Unexpected character '@' (1:0)

### ISSUE THREAD

**Issue by rahch1982** [vuetifyjs/cordova#22]
I'm working on a new project using this template through vue-cli. I used the exact commands provided in this repo to install the template. On a fresh installation, executing 'npm run dev' or 'npm run build' works just fine. After upgrading to **Webpack 4.8.3** and **vue-loader 15.2.0**; and after patching 'build/webpack.base.conf.js', 'build/utils.js', and 'build/webpack.prod.conf.js' to replace **'extract-text-webpack-plugin'** for **'mini-css-extract-plugin'** (plugin supported on webpack 4, meant to replace extract-text plugin and provide you with better future compatibility), executing  'npm run build' throws the following error:

> ERROR in ./node_modules/vuetify/dist/vuetify.css
Module build failed: ModuleBuildError: Module build failed: ModuleParseError: Module parse failed: Unexpected character '@' (1:0)
You may need an appropriate loader to handle this file type.
| @-webkit-keyframes shake {
|   59% {
|     margin-left: 0;
   ...
 @ ./src/main.js 5:0-34
 @ multi babel-polyfill ./src/main.js

> ERROR in chunk app [entry]

I've made the following changes to the files mentioned above:

## **build/webpack.base.conf.js** 
```
// var ExtractTextPlugin = require('extract-text-webpack-plugin')
var MiniCssExtractPlugin = require("mini-css-extract-plugin")
...
module.exports = {
...
optimization: {
    splitChunks: {
      name: true,
      chunks: 'all',
      cacheGroups: {
        vendors: {
          test: /node_modules/,
          enforce: true
        }
      }
    },
    minimize: !devMode ? true : false
  },
...
module: {
rules: [
...
{
          test: /\.(scss|css)$/,
          use: [
            MiniCssExtractPlugin.loader,
            {
              loader: 'css-loader',
              options: {
                minimize: {
                  safe: true
                },
                sourceMap: true
              }
            },
            {
              loader: 'postcss-loader',
              options: {
                autoprefixer: {
                  browsers: ['last 2 versions']
                },
                plugins: () => [
                  autoprefixer
                ],
                sourceMap: true
              },
            },
            {
              loader: 'sass-loader',
              options: {
                sourceMap: true
              }
            }
          ]
  }
...
```

## **build/webpack.prod.conf.js** 

```
...
// var ExtractTextPlugin = require('extract-text-webpack-plugin')
var MiniCssExtractPlugin = require("mini-css-extract-plugin")
...
var webpackConfig = merge(baseWebpackConfig, {
...
plugins: [
    ...
    // extract css into its own file
    // new ExtractTextPlugin({
    //   filename: utils.assetsPath('css/[name].[md5:contenthash:hex:15].css')
    // }),
    new MiniCssExtractPlugin({
      filename: utils.assetsPath('[name].[hash].css'),
      chunkFilename: utils.assetsPath('[id].[hash].css')
    }),
    ...
],
...
```

## build/utils.js

```
...
// var ExtractTextPlugin = require('extract-text-webpack-plugin')
var MiniCssExtractPlugin = require('mini-css-extract-plugin')
...
// generate loader string to be used with extract text plugin
  function generateLoaders (loader, loaderOptions) {
    ...
    // Extract CSS when that option is specified
    // (which is the case during production build)
    if (options.extract) {
      //return ExtractTextPlugin.extract({
      //  use: loaders,
      //  fallback: 'vue-style-loader'
      //})
      return MiniCssExtractPlugin.loader
    ...
```

Any hints on what could I've miss to make this working? I would appreciate your attention. Thanks in advanced.

**Comment by cairinmichie** [comment]
Do you have a reason for upgrading webpack and vue-loader?

If not don't upgrade them. I don't have much experience with webpack, so I can't really help you here.

**Comment by rahch1982** [comment]
Hello @cairinmichie , thanks for the fast answer. I'm upgrading to the latest version of webpack and vue-loader, to take advantage of the changes committed to those modules.

I finally got it working. The issue is directly related to **vue-loader (15.2.0)**. After making some tests on a newly fresh installation of this template, I noticed that module "mini-css-extract-plugin" was not working as expected with vue-loader version 15.2.0.

I installed module **"extract-text-webpack-plugin"** version **4.0.0-beta.0** ; and patched '**build/webpack.base.conf.js**', '**build/utils.js**', and '**build/webpack.prod.conf.js**'. After making those changes, executing 'npm run dev' or 'npm run build' works as expected.

In case this can help anyone, below you can find all changes to the files mentioned above:

## build/webpack.base.conf.js

```
var path = require('path')
var utils = require('./utils')
var config = require('../config')
var vueLoaderConfig = require('./vue-loader.conf')

const { VueLoaderPlugin } = require('vue-loader')

function resolve (dir) {
  return path.join(__dirname, '..', dir)
}

module.exports = {
  entry: {
    app: ['babel-polyfill','./src/main.js']
  },
  output: {
    path: config.build.assetsRoot,
    filename: '[name].js',
    publicPath: process.env.NODE_ENV === 'production'
      ? config.build.assetsPublicPath
      : config.dev.assetsPublicPath
  },
  resolve: {
    extensions: ['.js', '.vue', '.json'],
    alias: {
      'vue$': 'vue/dist/vue.esm.js',
      '@': resolve('src')
    }
  },
  module: {
    rules: [
      {
        test: /\.(js|vue)$/,
        loader: 'eslint-loader',
        enforce: 'pre',
        include: [resolve('src'), resolve('test')],
        options: {
          formatter: require('eslint-friendly-formatter')
        }
      },
      {
        test: /\.vue$/,
        loader: 'vue-loader',
        options: vueLoaderConfig
      },
      {
        test: /\.js$/,
        loader: 'babel-loader',
        include: [resolve('src'), resolve('test')]
      },
      {
        test: /\.(png|jpe?g|gif|svg)(\?.*)?$/,
        loader: 'url-loader',
        options: {
          limit: 10000,
          name: utils.assetsPath('img/[name].[hash:7].[ext]')
        }
      },
      {
        test: /\.(mp4|webm|ogg|mp3|wav|flac|aac)(\?.*)?$/,
        loader: 'url-loader',
        options: {
          limit: 10000,
          name: utils.assetsPath('media/[name].[hash:7].[ext]')
        }
      },
      {
        test: /\.(woff2?|eot|ttf|otf)(\?.*)?$/,
        loader: 'url-loader',
        options: {
          limit: 10000,
          name: utils.assetsPath('fonts/[name].[hash:7].[ext]')
        }
      }
    ]
  },
  plugins: [
    new VueLoaderPlugin()
  ]
}
```

## build/webpack.prod.conf.js

```
var fs = require('fs')
var path = require('path')
var utils = require('./utils')
var webpack = require('webpack')
var config = require('../config')
var merge = require('webpack-merge')
var baseWebpackConfig = require('./webpack.base.conf')
var CopyWebpackPlugin = require('copy-webpack-plugin')
var HtmlWebpackPlugin = require('html-webpack-plugin')
var ExtractTextPlugin = require('extract-text-webpack-plugin')
var OptimizeCSSPlugin = require('optimize-css-assets-webpack-plugin')
var SWPrecacheWebpackPlugin = require('sw-precache-webpack-plugin')
var loadMinified = require('./load-minified')
var UglifyJsPlugin = require('uglifyjs-webpack-plugin')

var env = process.env.NODE_ENV === 'testing'
  ? require('../config/test.env')
  : config.build.env

var webpackConfig = merge(baseWebpackConfig, {
  mode: 'production',
  module: {
    rules: utils.styleLoaders({
      sourceMap: config.build.productionSourceMap,
      extract: true
    })
  },
  devtool: config.build.productionSourceMap ? '#source-map' : false,
  output: {
    path: config.build.assetsRoot,
    filename: utils.assetsPath('js/[name].[chunkhash:15].js'),
    chunkFilename: utils.assetsPath('js/[id].[chunkhash:15].js')
  },
   optimization: {
    minimizer: [
      // we specify a custom UglifyJsPlugin here to get source maps in production
      new UglifyJsPlugin({
        cache: true,
        parallel: true,
        uglifyOptions: {
          compress: false,
          ecma: 6,
          mangle: true
        },
        sourceMap: true
      })
    ]
  },
  plugins: [
    new webpack.LoaderOptionsPlugin({ options: {} }),
    // http://vuejs.github.io/vue-loader/en/workflow/production.html
    new webpack.DefinePlugin({
      'process.env': env
    }),
    // extract css into its own file
    new ExtractTextPlugin({
      filename: utils.assetsPath('css/[name].[md5:contenthash:hex:15].css')
    }),
    // Compress extracted CSS. We are using this plugin so that possible
    // duplicated CSS from different components can be deduped.
    new OptimizeCSSPlugin({
      cssProcessorOptions: {
        safe: true
      }
    }),
    // generate dist index.html with correct asset hash for caching.
    // you can customize output by editing /index.html
    // see https://github.com/ampedandwired/html-webpack-plugin
    new HtmlWebpackPlugin({
      filename: process.env.NODE_ENV === 'testing'
        ? 'index.html'
        : config.build.index,
      template: 'index.html',
      inject: true,
      minify: {
        removeComments: true,
        collapseWhitespace: true,
        removeAttributeQuotes: true
        // more options:
        // https://github.com/kangax/html-minifier#options-quick-reference
      },
      // necessary to consistently work with multiple chunks via CommonsChunkPlugin
      chunksSortMode: 'dependency',
      serviceWorkerLoader: `<script>${loadMinified(path.join(__dirname,
        './service-worker-prod.js'))}</script>`
    }),
    // copy custom static assets
    new CopyWebpackPlugin([
      {
        from: path.resolve(__dirname, '../static'),
        to: config.build.assetsSubDirectory,
        ignore: ['.*']
      }
    ]),
    // service worker caching
    new SWPrecacheWebpackPlugin({
      cacheId: 'my-vue-app',
      filename: 'service-worker.js',
      staticFileGlobs: ['dist/**/*.{js,html,css}'],
      minify: true,
      stripPrefix: 'dist/'
    })
  ]
})

if (config.build.productionGzip) {
  var CompressionWebpackPlugin = require('compression-webpack-plugin')

  webpackConfig.plugins.push(
    new CompressionWebpackPlugin({
      asset: '[path].gz[query]',
      algorithm: 'gzip',
      test: new RegExp(
        '\\.(' +
        config.build.productionGzipExtensions.join('|') +
        ')$'
      ),
      threshold: 10240,
      minRatio: 0.8
    })
  )
}

if (config.build.bundleAnalyzerReport) {
  var BundleAnalyzerPlugin = require('webpack-bundle-analyzer').BundleAnalyzerPlugin
  webpackConfig.plugins.push(new BundleAnalyzerPlugin())
}

module.exports = webpackConfig
```

## build/utils.js

```
var path = require('path')
var config = require('../config')
var ExtractTextPlugin = require('extract-text-webpack-plugin')

exports.assetsPath = function (_path) {
  var assetsSubDirectory = process.env.NODE_ENV === 'production'
    ? config.build.assetsSubDirectory
    : config.dev.assetsSubDirectory
  return path.posix.join(assetsSubDirectory, _path)
}

exports.cssLoaders = function (options) {
  options = options || {}

  var cssLoader = {
    loader: 'css-loader',
    options: {
      minimize: process.env.NODE_ENV === 'production',
      sourceMap: options.sourceMap
    }
  }

  // generate loader string to be used with extract text plugin
  function generateLoaders (loader, loaderOptions) {
    var loaders = [cssLoader]
    if (loader) {
      loaders.push({
        loader: loader + '-loader',
        options: Object.assign({}, loaderOptions, {
          sourceMap: options.sourceMap
        })
      })
    }

    // Extract CSS when that option is specified
    // (which is the case during production build)
    if (options.extract) {
      return ExtractTextPlugin.extract({
        use: loaders,
        fallback: 'vue-style-loader'
      })
    } else {
      return ['vue-style-loader'].concat(loaders)
    }
  }

  // https://vue-loader.vuejs.org/en/configurations/extract-css.html
  return {
    css: generateLoaders(),
    postcss: generateLoaders(),
    less: generateLoaders('less'),
    sass: generateLoaders('sass', { indentedSyntax: true }),
    scss: generateLoaders('sass'),
    stylus: generateLoaders('stylus'),
    styl: generateLoaders('stylus')
  }
}

// Generate loaders for standalone style files (outside of .vue)
exports.styleLoaders = function (options) {
  var output = []
  var loaders = exports.cssLoaders(options)
  for (var extension in loaders) {
    var loader = loaders[extension]
    output.push({
      test: new RegExp('\\.' + extension + '$'),
      use: loader
    })
  }
  return output
}
```

**Comment by cairinmichie** [comment]
Great work! If you want you can make a PR, otherwise I will only be able to update this on the weekend.

**Comment by rahch1982** [comment]
Hello @cairinmichie , I've made the PR as requested. Let me know if there is anything else I can do in order to contribute.

**Comment by rahch1982** [comment]
Hello @cairinmichie , it's been a while!. After a few months, I could successfully replace the deprecated module "extract-text-webpack-plugin" to 'mini-css-extract-plugin' and all its dependences.  I have made a few changes to the files mentioned above and now, the execution of 'npm run dev' or 'npm run build' works as expected.

I've created a pull request that will remove deprecated modules from the template, and will include those packages that will provide compatibility with newer versions of Webpack and Vue-loader. Let me know if there is anything else I can do in order to keep  contributing.
